src/librustc_borrowck/borrowck/mir/elaborate_drops.rs

```diff
@@ -0,0 +1,1094 @@
+                // dead drop - patch it out
+                self.patch.patch_terminator(bb, TerminatorKind::Goto {
+                    target: c.succ
+                });
+            }
+            (true, false) => {
+                // static drop - just set the flag
+                self.patch.patch_terminator(bb, TerminatorKind::Drop {
+                    location: c.lvalue.clone(),
+                    target: c.succ,
+                    unwind: c.unwind
+                });
+                self.drop_flags_for_drop(c, bb);
+            }
+            (true, true) => {
+                // dynamic drop
+                if children_count == 1 || self.must_complete_drop(c) {
+                    self.conditional_drop(c, bb);
+                } else {
+                    self.partial_drop(c, bb);
+                }
+            }
+        }
+    }
+
+    fn move_paths_for_fields(&self,
+                             base_lv: Lvalue<'tcx>,
+                             variant_path: MovePathIndex,
+                             variant: ty::VariantDef<'tcx>,
+                             substs: &'tcx Substs<'tcx>)
+                             -> Vec<(Lvalue<'tcx>, Option<MovePathIndex>)>
+    {
+        let move_paths = &self.move_data().move_paths;
+        variant.fields.iter().enumerate().map(|(i, f)| {
+            let subpath =
+                super::move_path_children_matching(move_paths, variant_path, |p| {
+                    match p {
+                        &Projection {
+                            elem: ProjectionElem::Field(idx, _), ..
+                        } => idx.index() == i,
+                        _ => false
+                    }
+                });
+
+            let field_ty =
+                self.tcx().normalize_associated_type_in_env(
+                    &f.ty(self.tcx(), substs),
+                    &self.param_env
+                );
+            (base_lv.clone().field(Field::new(i), field_ty), subpath)
+        }).collect()
+    }
+
+    fn drop_ladder<'a>(&mut self,
+                       c: &DropCtxt<'a, 'tcx>,
+                       unwind_ladder: Option<Vec<BasicBlock>>,
+                       succ: BasicBlock,
+                       fields: &[(Lvalue<'tcx>, Option<MovePathIndex>)],
+                       is_cleanup: bool)
+                       -> Vec<BasicBlock>
+    {
+        let mut succ = succ;
+        let mut unwind_succ = if is_cleanup {
+            None
+        } else {
+            c.unwind
+        };
+        let mut seen_self = false;
+
+        fields.iter().rev().enumerate().map(|(i, &(ref lv, path))| {
+            let field_c = DropCtxt {
+                span: c.span,
+                scope: c.scope,
+                is_cleanup: is_cleanup,
+                init_data: c.init_data,
+                kind: if path.is_some() {
+                    ElaborateKind::Normal
+                } else if seen_self {
+                    ElaborateKind::RestHead
```

Comment:
these cases are subtle, could you add some comments specifically about why you are using each mode here?